### PR TITLE
Start preservation command

### PR DESCRIPTION
**Pull request by janid1967**
StartPreservationCommand, validator and handler
Update Requirement.NeedsUserInput from RequirementDefinition when starting preservation
Refactor Validators to use CascadeMode.StopOnFirstFailure. Works within each chain of RuleFor. Idea is to for instance don't check for IsVoided  if don't exists
Extension on DateTime to format DateTime as week and year

### CODE REVIEW

File: src/Equinor.Procosys.Preservation.Command/TagCommands/CreateTag/CreateTagCommand.cs [hunk 1]
```diff
@@ -11,7 +12,7 @@ public CreateTagCommand(string tagNo, string projectNo, int stepId, IEnumerable<
             TagNo = tagNo;
             ProjectNo = projectNo;
             StepId = stepId;
-            Requirements = requirements;
+            Requirements = requirements ?? throw new ArgumentNullException(nameof(requirements));
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/TagCommands/CreateTag/CreateTagCommand.cs]: Should there be input checking in the commands? There's already input validation on the DTO and there is more validation in the command validator. What should and shouldn't be allowed here?

File: src/Equinor.Procosys.Preservation.Command/TagCommands/StartPreservation/StartPreservationCommand.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+using System.Collections.Generic;
+using MediatR;
+using ServiceResult;
+
+namespace Equinor.Procosys.Preservation.Command.TagCommands.StartPreservation
+{
+    public class StartPreservationCommand : IRequest<Result<Unit>>
+    {
+        public StartPreservationCommand(IEnumerable<int> tagIds)
+            => TagIds = tagIds ?? throw new ArgumentNullException(nameof(tagIds));
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/TagCommands/StartPreservation/StartPreservationCommand.cs]: Ref. earlier comment regarding argument validation.

File: src/Equinor.Procosys.Preservation.Command/Validators/Tag/ITagValidator.cs [hunk 3]
```diff
@@ -9,5 +11,11 @@ public interface ITagValidator
         bool IsVoided(int tagId);
         
         bool ProjectIsClosed(int tagId);
+        
+        bool VerifyPreservationStatus(int tagId, PreservationStatus status);
+        
+        bool HasANonVoidedRequirement(int tagId);
+        
+        bool AllRequirementDefinitionsExists(int tagId);
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/Validators/Tag/ITagValidator.cs]: AllRequirementDefinitionsExists -> AllRequirementDefinitionsExist.

File: src/Equinor.Procosys.Preservation.Command/TagCommands/CreateTag/CreateTagCommand.cs (hunk 1 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Command/TagCommands/CreateTag/CreateTagCommand.cs]: Changed to newing empty list

File: src/Equinor.Procosys.Preservation.Command/TagCommands/StartPreservation/StartPreservationCommand.cs (hunk 2 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Command/TagCommands/StartPreservation/StartPreservationCommand.cs]: Changed to newing empty list

File: src/Equinor.Procosys.Preservation.Command/Validators/Tag/TagValidator.cs [hunk 4]
```diff
@@ -22,5 +28,30 @@ public bool IsVoided(int tagId)
...
+        public bool HasANonVoidedRequirement(int tagId)
+        {
+            var tag = _tagRepository.GetByIdAsync(tagId).Result;
+            return tag?.Requirements != null && tag.Requirements.Any(r => !r.IsVoided);
+        }
+
+        public bool AllRequirementDefinitionsExists(int tagId)
+        {
+            var tag = _tagRepository.GetByIdAsync(tagId).Result;
+            if (tag == null || tag.Requirements == null)
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/Validators/Tag/TagValidator.cs]: Can be simplified to if (tag?.Requirements == null)

File: src/Equinor.Procosys.Preservation.Domain/PreservationDate.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+        private int _week;
+
+        public PreservationDate(DateTime time)
+        {
+            _time = time;
+            _year = ISOWeek.GetYear(time);
+            _week = ISOWeek.GetWeekOfYear(time);
+        }
+
+        public override string ToString() => string.Concat(_year.ToString(), "W", _week.ToString("00"));
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Domain/PreservationDate.cs]: Use lower case 'w'.

File: src/Equinor.Procosys.Preservation.Domain/PreservationDate.cs [hunk 6]
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.Globalization;
+
+namespace Equinor.Procosys.Preservation.Domain
+{
+    public class PreservationDate
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Domain/PreservationDate.cs]: Consider making a DateTime extension method "FormatAsYearAndWeekString(this DateTime dateTime)" or similar instead.

File: src/Equinor.Procosys.Preservation.Query/TagAggregate/RequirementDto.cs [hunk 7]
```diff
@@ -1,18 +1,22 @@
-﻿namespace Equinor.Procosys.Preservation.Query.TagAggregate
+﻿using System;
+using Equinor.Procosys.Preservation.Domain;
+
+namespace Equinor.Procosys.Preservation.Query.TagAggregate
 {
     public class RequirementDto
     {
-        public RequirementDto(int id, int requirementDefinitionId, bool isVoided, int intervalWeeks)
+        private readonly DateTime? _nextDueTimeUtc;
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Query/TagAggregate/RequirementDto.cs]: Make this a public property to make the DTO more generic.

File: src/Equinor.Procosys.Preservation.Domain/PreservationDate.cs (hunk 6 above)
Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Domain/PreservationDate.cs]: Made extension

File: src/Equinor.Procosys.Preservation.Command/TagCommands/StartPreservation/StartPreservationCommandHandler.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            foreach (var tag in tags)
+            {
+                var reqDefs = await _requirementTypeRepository
+                    .GetRequirementDefinitionsByIdsAsync(tag.Requirements.Select(r => r.RequirementDefinitionId)
+                        .ToList());
+                foreach (var requirement in tag.Requirements)
+                {
+                    var reqDef = reqDefs.Single(r => r.Id == requirement.RequirementDefinitionId);
+
+                    requirement.StartPreservation(_timeService.GetCurrentTimeUtc(), reqDef.NeedsUserInput);
```
Comment (hbjorgo) [inline src/Equinor.Procosys.Preservation.Command/TagCommands/StartPreservation/StartPreservationCommandHandler.cs]: Move this call to Tag.StartPreservation(...).

Comment (janid1967) [reply src/Equinor.Procosys.Preservation.Command/TagCommands/StartPreservation/StartPreservationCommandHandler.cs]: NeedsUserInput removed. Will be implemented on Tag later. Tag.NeedsUserInput set to false for iteration 1

**Pull request merged**